cryptopan/src/cryptopan.rs

```diff
@@ -0,0 +1,225 @@
+//! `CryptoPAn`
+//! IP-address anonymization algorithm that preserves prefixes
+//! Tightly based on the GO implementation
+//! Author: Yawning Angel <yawning at schwanenlied dot me>
+//! [Github Link](./https://github.com/Yawning/cryptopan/blob/master/cryptopan.go)
+//! Based upon the algorithm described in:
+//! Package cryptopan implements the Crypto-PAn prefix-preserving IP address sanitization algorithm as specified by J. Fan, J. Xu, M. Ammar, and S. Moon.
+use aes::cipher::BlockEncryptMut;
+use aes::cipher::{generic_array::GenericArray, KeyInit};
+use bitvec::field::BitField;
+use bitvec::prelude::Msb0;
+use bitvec::{bitvec, vec::BitVec};
+use std::net::{
+    IpAddr,
+    IpAddr::{V4, V6},
+    Ipv4Addr, Ipv6Addr,
+};
+
+type Aes128EcbEnc = ecb::Encryptor<aes::Aes128>;
+
+pub struct CryptoPAn {
+    aes: Aes128EcbEnc,
+    pad: BitVec<u8, Msb0>,
+}
+
+impl CryptoPAn {
+    /// Creates new outer block for encryption using key. Panics if key size isn't 256 bits
+    pub fn from_key(key: &[u8]) -> Self {
+        // Split AES key into two parts - one is padding
+        // Encrypt the padding with the first part of the key
+        let mut aes = Aes128EcbEnc::new(GenericArray::from_slice(&key[0..16]));
+
+        let mut buf = [0u8; 16];
+        buf.copy_from_slice(&key[16..32]);
+        let g = GenericArray::from_mut_slice(&mut buf);
+        aes.encrypt_block_mut(g);
+
+        Self {
+            aes,
+            pad: BitVec::from_slice(g.as_slice()),
+        }
+    }
+    /// Creates new outer block for encryption using str key. Panics if key size isn't 256 bits.
+    pub fn from_str_key(string_key: &str) -> Self {
+        Self::from_key(string_key.as_bytes())
+    }
+    /// Encrypts either a V4 or V6 IP address using the key
+    pub fn encrypt(&self, input: IpAddr) -> IpAddr {
```

Comment:
I think it will be better to define separate `encrypt_v4` and `encrypt_v6` methods, making `encrypt` a simple matching wrapper around them. Also I think we can do without `bitvec`.

Reply:
Would you just shift bits to do the operations instead of using `bitvec`? I think using `bitvec` is cleaner and easier to understand, but I understand replacing it if it performs more poorly than working with the bits directly. Or perhaps there is another reason not to use it?